### PR TITLE
More inertia fixes

### PR DESCRIPTION
**Pull request by ebianchi**
The [previous PR with inertial parameter fixes](https://github.com/DAIRLab/dair_pll/pull/2) did not sufficiently address the disparity in how inertial parameters in `pi` format require moments of inertia w.r.t. centers of mass, while inertial parameters in `theta` format require moments of inertia w.r.t. body origins.  That previous PR managed to load inertial parameters in a way that made the original `pi` parameters match those provided in the URDF.  However, it neglected to convert to/from `theta` format accurately, given `pi` format did not contain the correct moments of inertia expected by the Rucker and Wensing `pi`-to-`theta` conversion.

To address this discrepancy, this PR proposes to replace any mention of `pi` parameters with one of two alternatives:  `pi_o` or `pi_cm`, both in the original `pi` format but with moments of inertia expressed w.r.t. the body origin or the body center of mass, respectively.  Inertial parameters can be converted between these two via `InertialParameterConverter.pi_o_to_pi_cm()` and `InertialParameterConverter.pi_cm_to_pi_o()`, which perform the conversions using the Parallel Axis Theorem.

Thus, `theta` format is calculated via Rucker and Wensing's method, starting from `pi_o` format.

This PR primarily involves the additional necessary functionality in `inertia.py` to represent and convert to/from the new inertial parameter formats.  Secondary files `multibody_terms.py` and `urdf_utils.py` feature variable name changes to clarify `pi_o` versus `pi_cm` format and function call changes to reference new methods in `inertia.py`.

### CODE REVIEW

**Comment by ebianchi** [comment]
The latest commit ([e470600](https://github.com/DAIRLab/dair_pll/pull/4/commits/e470600b764f41fd877187a3b627a7d7bc00d646)) corrected a bug in the `pi_o` to `pi_cm` and `pi_cm` to `pi_o` conversions which turned positive any negative `I_xy, I_xz, I_yz` off-diagonal moments of inertia.  None of my original test systems matched this scenario, so I had missed it earlier.  Thanks to some more extensive testing and @mposa for identifying this behavior.

File: dair_pll/inertia.py [hunk 1]
```diff
@@ -99,18 +106,96 @@ def number_to_float(number: Any) -> float:
     return float(str(number))
 
 
+def parallel_axis_theorem(inertia_mat: Tensor, masses: Tensor, vec: Tensor,
```
Comment (mshalm) [inline dair_pll/inertia.py]: Well documented and implemented. Some stylistic suggestions:

- Change arguments to drake formate. i.e. `inertia_mat` -> `I_BBa_B`, `masses` -> `m_B`, vec -> `p_BaBb_B`, `from_com_to_other` -> `Ba_is_Bcm`, where we note from the squaring that the sign on the vector doesn't matter, i.e. you could also give `p_BbBa_B`
- Double check the Sphinx autodoc of the comment -- some things are code-formatted when they should be math-formatted

File: dair_pll/inertia.py [hunk 2]
```diff
@@ -99,18 +106,96 @@ def number_to_float(number: Any) -> float:
...
+    d_squared = skew_symmetric(vec) @ skew_symmetric(vec)
+    term = d_squared * masses.view((-1, 1, 1))
+
+    if from_com_to_other:
+        return inertia_mat - term
+    else:
+        return inertia_mat + term
+
+
+def inertia_matrix_from_vector(inertia_vec: Tensor):
```
Comment (mshalm) [inline dair_pll/inertia.py]: suggest drake-formatted variable names

File: dair_pll/inertia.py [hunk 3]
```diff
@@ -99,18 +106,96 @@ def number_to_float(number: Any) -> float:
...
+    # Put Ixx, Iyy, Izz on the diagonals.
+    diags = torch.diag_embed(inertia_vec[:, :3])
+
+    # Put Ixy, Ixz, Iyz on the off-diagonals.
+    off_diags = nonskew_symmetric(inertia_vec[:, 3:].flip(1))
+
+    return diags + off_diags
+
+
+def inertia_vector_from_matrix(inertia_mat: Tensor):
```
Comment (mshalm) [inline dair_pll/inertia.py]: suggest drake-formatted variable names

File: dair_pll/inertia.py [hunk 4]
```diff
@@ -203,14 +288,82 @@ def pi_to_theta(pi: Tensor) -> Tensor:
         return torch.cat((alpha, d_vector, s_and_t), -1)
 
     @staticmethod
-    def pi_to_drake_spatial_inertia(pi: Tensor) -> Tensor:
+    def pi_o_to_pi_cm(pi_o: Tensor) -> Tensor:
```
Comment (mshalm) [inline dair_pll/inertia.py:305]: originating from some of my poor prior documentation, I'm noticing that `pi_o_to_theta` can handle arbitrary dimensions of batching, whereas it's comments suggest it's only being used with 1 batching dimension. This function can also only handle 1 dimension. Suggest adding a `assert pi_o.dim() <= 2` s to make sure that the function doesn't get called with e.g. a `(*,*,10)`

File: dair_pll/inertia.py [hunk 5]
```diff
@@ -203,14 +288,82 @@ def pi_to_theta(pi: Tensor) -> Tensor:
...
+        # Use parallel axis theorem to compute inertia matrix wrt CoM.
+        inertia_mat = inertia_matrix_from_vector(I_BBo_B)
+        new_inertia_mat = parallel_axis_theorem(inertia_mat, mass, p_BoBcm_B,
+                                                from_com_to_other=False)
+        I_BBcm_B = inertia_vector_from_matrix(new_inertia_mat)
+
+        return torch.hstack((mass, p_BoBcm_B*mass, I_BBcm_B)).reshape(-1, 10)
+
+    @staticmethod
+    def pi_cm_to_pi_o(pi_cm: Tensor) -> Tensor:
```
Comment (mshalm) [inline dair_pll/inertia.py:334]: `assert` on dimensionality

File: dair_pll/inertia.py [hunk 6]
```diff
@@ -203,14 +288,82 @@ def pi_to_theta(pi: Tensor) -> Tensor:
...
+    @staticmethod
+    def pi_cm_to_theta(pi_cm: Tensor) -> Tensor:
+        """Passthrough chain of :py:meth:`pi_cm_to_pi_o` and
+        :py:meth:`pi_o_to_theta`."""
+        pi_o = InertialParameterConverter.pi_cm_to_pi_o(pi_cm)
+        return InertialParameterConverter.pi_o_to_theta(pi_o)
+
+    @staticmethod
+    def pi_cm_to_drake_spatial_inertia(pi: Tensor) -> Tensor:
         """Converts batch of ``pi`` parameters to ``drake_spatial_inertia``
```
Comment (mshalm) [inline dair_pll/inertia.py]: should be `pi_cm` in comment

File: dair_pll/inertia.py [hunk 7]
```diff
@@ -237,22 +390,41 @@ def drake_to_pi(spatial_inertia: DrakeSpatialInertia) -> Tensor:
             number_to_float(I_BBcm_B[index[0], index[1]])
             for index in INERTIA_INDICES
         ]
-        pi = Tensor([mass] + mass_list + inertia_list)
-        return pi
+        pi_cm = Tensor([mass] + mass_list + inertia_list)
+        return pi_cm
+
+    @staticmethod
+    def drake_to_pi_o(spatial_inertia: DrakeSpatialInertia) -> Tensor:
```
Comment (mshalm) [inline dair_pll/inertia.py:408]: is there a reason this isn't implemented as a passthrough like `pi_cm_to_theta`?

File: dair_pll/inertia.py [hunk 8]
```diff
@@ -237,22 +390,41 @@ def drake_to_pi(spatial_inertia: DrakeSpatialInertia) -> Tensor:
...
+        :py:attr:`~pydrake.multibody.tree.SpatialInertia` object."""
+        mass = number_to_float(spatial_inertia.get_mass())
+        p_BoBcm_B = spatial_inertia.get_com()
+        I_BBo_B = spatial_inertia.CalcRotationalInertia()
+
+        mass_list = [
+            mass * number_to_float(coordinate) for coordinate in p_BoBcm_B
+        ]
+
+        inertia_list = [
```
Comment (mshalm) [inline dair_pll/inertia.py:428]: Could this be made more compact using your new inertia vector <--> matrix conversions?

File: dair_pll/multibody_terms.py [hunk 9]
```diff
@@ -534,8 +534,9 @@ def scalars_and_meshes(
                 self.plant_diagram.plant,
                 self.plant_diagram.model_ids)
 
-        for body_pi, body_id in zip(self.lagrangian_terms.pi(), all_body_ids):
-            body_scalars = InertialParameterConverter.pi_to_scalars(body_pi)
+        for body_pi, body_id in zip(self.lagrangian_terms.pi_cm(), all_body_ids):
```
Comment (mshalm) [inline dair_pll/multibody_terms.py:541]: calls a `pi_cm` method on the `LagrangianTerms` class which doesn't exist in the commit (likely a renaming of `LagrangianTerms.pi(self)`)

File: dair_pll/tensor_utils.py [hunk 10]
```diff
@@ -162,6 +162,35 @@ def skew_symmetric(vectors: Tensor) -> Tensor:
     return torch.stack((row_1, row_2, row_3), -2)
 
 
+def nonskew_symmetric(vectors: Tensor) -> Tensor:
```
Comment (mshalm) [inline dair_pll/tensor_utils.py]: Naming is a little clunky. The space of skew symmetric matrices has a 3-dimensional basis, so the conversion in that case is a little more clear; also "nonskew symemtric" seems to just suggest "symmetric" and leaves out "zero diagonal". Suggest refactor to `symmetric_offdiagonal` or something similar

**Review by mshalm — CHANGES_REQUESTED** [review]
Overall styled and implemented well. A couple apparent bugs/weak points (`LagrangianTerms.pi` and some dimensional `assert` s ) that should be examined, and some non-required refactoring and documentation suggestions.

File: dair_pll/multibody_terms.py (hunk 9 above)
Comment (ebianchi) [reply dair_pll/multibody_terms.py:541]: Good catch -- your guess is correct.  Propagated the name change with [this commit](https://github.com/DAIRLab/dair_pll/pull/4/commits/b1faed21be2e84611ab8d337f99960f9f76d9c4f).

File: dair_pll/tensor_utils.py (hunk 10 above)
Comment (ebianchi) [reply dair_pll/tensor_utils.py]: Great suggestion -- I adopted exactly that suggested name and updated the documentation slightly in [this commit](https://github.com/DAIRLab/dair_pll/pull/4/commits/8db8ec87cbf382d33a38f32d077391f8e47e125d).

File: dair_pll/inertia.py (hunk 6 above)
Comment (ebianchi) [reply dair_pll/inertia.py]: There were a few other instances where `pi` could be distinguished as either `pi_o` or `pi_cm`.  All addressed in [this commit](https://github.com/DAIRLab/dair_pll/pull/4/commits/ae513f85f7ae23ff97ba48a87b46f8d69efee029).

File: dair_pll/inertia.py (hunk 1 above)
Comment (ebianchi) [reply dair_pll/inertia.py]: When correcting the Sphinx autodoc of `inertia.py`, I noticed some other minor but broken Sphinx-related documentation issues.  Without changing the documentation, [this commit](https://github.com/DAIRLab/dair_pll/pull/4/commits/07741f51953bb16b79f7582af7494b2571989f30) also corrected any of those that were improperly formatted.  FYI when you see there are more files with edits than initially in this PR.

File: dair_pll/inertia.py (hunk 7 above)
Comment (ebianchi) [reply dair_pll/inertia.py:408]: You're correct in that the current implementation relying on Drake's `CalcRotationalInertia` and `Shift` (for `drake_to_pi_cm`) is equivalent to a pass-through, e.g. `drake_to_pi_o(*) = pi_cm_to_pi_o(drake_to_pi_cm(*))`.  Since either is technically correct, I chose to rely on Drake's infrastructure rather than ours just since theirs is maintained by a team and might be less error-prone and more efficient.  However I'm happy to switch to a pass-through approach if you have reasons to prefer it, even if just stylistic.

File: dair_pll/inertia.py (hunk 8 above)
Comment (ebianchi) [reply dair_pll/inertia.py:428]: It seemed like this could almost work.  One can do:
```
I_BBo_B_mat = I_BBo_B.CopyToFullMatrix3()
```
That creates a numpy array of the inertia matrix, however of `pydrake.symbolic.Expression` objects with `dtype=object`.  It's possible there's a way to convert an entire numpy array full of `pydrake.symbolic.Expression`s from `dtype=object` to `dtype=np.float64`, however I haven't found one yet.  Thus any solution that gets the inertia matrix via `RotationalInertia.CopyToFullMatrix()` (or other options like `RotationalInertia.get_moments()` and `RotationalInertia.get_products()`) would still require an element-by-element conversion, similar to `drake_to_pi_cm`'s current implementation.

**Comment by ebianchi** [comment]
I addressed and/or responded to each of the comments.  The results of some lingering conversations include:

- `drake_to_pi_o`'s implementation relying on Drake's functionality instead of as a pass-through of other PLL methods:  let me know if you'd prefer the pass-through implementation instead.
- An exploration of simplifying the inertia parameter extraction from a `pydrake.multibody.tree.SpatialInertia` object:  After some exploration, I couldn't find a way that would avoid an element-by-element conversion like in the current implementation.
- Correcting minor Sphinx autodoc-related issues in other files:  While correcting the `inertia.py` autodocs, I took care of anything that triggered a warning and/or clearly was rendered in an unintentional way, without changing the contents of that other documentation.

File: dair_pll/inertia.py (hunk 7 above)
Comment (mshalm) [reply dair_pll/inertia.py:408]: The negative thing about the current implementation is that it's a copy-paste implementation, i.e. literally implemented by copying and pasting working code. Copy/paste tends to be bad because if we ever discover a bug with or want to change both functions, there's a risk of forgetting to do the same to the other function and there's also a doubled risk of making a typo in either one each time a modification is made; they also typically increase the total line-count of the file, such that there is more code to maintain.

A third option that minimizes copy/paste _and_ uses `CalcRotationalInertia` and `Shift` would be to implement a helper function `drake_inertial_components_to_pi(mass, p_BoBcm_B, I_BoBi_B)` that contains the code from `mass_list = ...` onward, and reimplement both `drake_to_pi_o` and `drake_to_pi_o` as 2-liners that call this function.

File: dair_pll/inertia.py (hunk 8 above)
Comment (mshalm) [reply dair_pll/inertia.py:428]: That's frustrating! seems like there isn't a good opportunity for improvement here then.

File: dair_pll/inertia.py (hunk 7 above)
Comment (ebianchi) [reply dair_pll/inertia.py:408]: I like that suggestion.  I'll do that now, thanks.

**Review by mshalm — APPROVED** [review]
LGTM
